src/python/gudhi/tensorflow/cubical_layer.py

```diff
@@ -0,0 +1,76 @@
+import numpy               as np
+import tensorflow          as tf
+from ..cubical_complex  import CubicalComplex
+
+######################
+# Cubical filtration #
+######################
+
+# The parameters of the model are the pixel values.
+
+def _Cubical(Xflat, Xdim, dimensions):
+    # Parameters: Xflat (flattened image),
+    #             Xdim (shape of non-flattened image)
+    #             dimensions (homology dimensions)
+
+    # Compute the persistence pairs with Gudhi
+    # We reverse the dimensions because CubicalComplex uses Fortran ordering
+    cc = CubicalComplex(dimensions=Xdim[::-1], top_dimensional_cells=Xflat)
+    cc.compute_persistence()
+
+    # Retrieve and ouput image indices/pixels corresponding to positive and negative simplices    
+    cof_pp = cc.cofaces_of_persistence_pairs()
+    
+    L_cofs = []
+    for dim in dimensions:
+
+        try:
+            cof = cof_pp[0][dim]
+        except IndexError:
+            cof = np.array([])
+
+        L_cofs.append(np.array(cof, dtype=np.int32))
+
+    return L_cofs
+
+class CubicalLayer(tf.keras.layers.Layer):
+    """
+    TensorFlow layer for computing the persistent homology of a cubical complex
+    """
+    def __init__(self, dimensions, min_persistence=None, **kwargs):
+        """
+        Constructor for the CubicalLayer class
+
+        Parameters:
+            dimensions (List[int]): homology dimensions
+            min_persistence (List[float]): minimum distance-to-diagonal of the points in the output persistence diagrams (default None, in which case 0. is used for all dimensions)
+        """
+        super().__init__(dynamic=True, **kwargs)
+        self.dimensions = dimensions
+        self.min_persistence = min_persistence if min_persistence != None else [0.] * len(self.dimensions)
+        assert len(self.min_persistence) == len(self.dimensions)
+
+    def call(self, X):
+        """
+        Compute persistence diagram associated to a cubical complex filtered by some pixel values 
+
+        Parameters:
+            X (TensorFlow variable): pixel values of the cubical complex
+
+        Returns:
+            dgms (list of TensorFlow variables): list of cubical persistence diagrams of length self.dimensions, where each element contains a finite persistence diagram of shape [num_finite_points, 2]
```

Comment:
I was going to say that this doesn't match what the other functions return, but actually the code outputs a list of arrays of shape (1,n,2). Is that meant as more or less equivalent to a list of tuples of arrays of shape (n,2) where the tuples have a single element? Anyway, that output is not obvious from the doc.

Reply:
@MathieuCarriere IIRC this is the only point blocking the merge of this branch, could you take a look?

Reply:
Now the documentation matches the code, that's good. On the other hand, the output for cubical is different from the output for Rips or lower-star, it requires a different number of `[0]` to extract the finite diagram.
* We can consider that this is not a problem, people won't often reuse the same code for the diagrams coming from Rips and Cubical, and even if they do the adaptation is not that bad. And since we don't output essential features for CubicalLayer it would be strange to leave an empty space for it.
* Or we can wrap the diagrams in a tuple of size 1.

That's something for you, @tlacombe and @hensel-f to decide, I am ok with both, I just want to make sure the choice is made consciously.